packages/node/src/utils/kyve/kyve.ts

```diff
@@ -0,0 +1,465 @@
+    }
+  }
+
+  private getBundleFromCache(height: number): BundleDetails | undefined {
+    return this.cachedBundleDetails.find(
+      (b) =>
+        parseDecimal(b.from_key) <= height && height <= parseDecimal(b.to_key),
+    );
+  }
+
+  private async updateCurrentBundleAndDetails(
+    height: number,
+  ): Promise<KyveBundleData[]> {
+    const bundle = this.getBundleFromCache(height);
+    if (bundle) {
+      return JSON.parse(await this.getFileCacheData(bundle));
+    } else {
+      const bundleId = await this.getBundleId(height);
+      const newBundleDetails = await this.getBundleById(bundleId);
+
+      this.addToCachedBundle(newBundleDetails);
+      return JSON.parse(await this.getFileCacheData(newBundleDetails));
+    }
+  }
+
+  private async pollUntilReadable(bundleFilePath: string): Promise<string> {
+    // eslint-disable-next-line no-constant-condition
+    while (true) {
+      try {
+        return await this.readFromFile(bundleFilePath);
+      } catch (e) {
+        if (e.code === 'EACCES') {
+          await delay(POLL_TIMER);
+        } else {
+          throw e;
+        }
+      }
+    }
+  }
+
+  async downloadAndProcessBundle(bundle: BundleDetails): Promise<void> {
+    const bundleFilePath = this.getBundleFilePath(bundle.id);
+
+    const writeStream = fs.createWriteStream(bundleFilePath, {
+      flags: 'wx',
+      mode: 0o200, // write only access for owner
+    });
+
+    // to ensure the stream to throw an on-stack error
+    await new Promise((resolve, reject) => {
+      writeStream.on('open', resolve);
+      writeStream.on('error', (err) => {
+        reject(err);
+      });
+    }).catch((e) => {
+      throw e;
+    });
+
+    const zippedBundleData = await this.retrieveBundleData(bundle.storage_id);
+
+    const gunzip = zlib.createUnzip({
+      maxOutputLength: MAX_COMPRESSION_BYTE_SIZE /* avoid zip bombs */,
+    });
+
+    await new Promise((resolve, reject) => {
+      zippedBundleData.data
+        .pipe(gunzip)
+        .on('error', reject)
+        .pipe(writeStream)
+        .on('finish', resolve);
+    }).catch((e) => {
+      throw e; // to ensure an on stack error is thrown
+    });
+
+    await fs.promises.chmod(bundleFilePath, 0o444);
+  }
+
+  private async getFileCacheData(bundle: BundleDetails): Promise<string> {
+    const bundleFilePath = this.getBundleFilePath(bundle.id);
+
+    try {
+      await this.downloadAndProcessBundle(bundle);
+      return await this.readFromFile(bundleFilePath);
+    } catch (e: any) {
+      if (['EEXIST', 'EACCES', 'ENOENT'].includes(e.code)) {
+        return this.pollUntilReadable(bundleFilePath);
+      } else {
+        if (axios.isAxiosError(e)) {
+          await fs.promises.unlink(bundleFilePath);
+        }
+
+        throw e;
+      }
+    }
+  }
+
+  async readFromFile(bundleFilePath: string): Promise<string> {
+    return fs.promises.readFile(bundleFilePath, 'utf-8');
+  }
+
+  private getToRemoveBundles(
```

Comment:
This can't rely on in memory data to determine data on disk. If there is a restart then the cache will be empty but there could be many bundle files.